feat: add model selector (fetch from v1/models) when user adds a provider model

lugnicca opened the pull request:
## Describe Your Changes

- users can either type custom model IDs (original way) or select from a dropdown of available models fetched from the provider's /v1/models endpoint
- automatically fetches models from provider APIs with caching
- full keyboard navigation with arrow keys, Enter, Escape, PageUp/PageDown
- more descriptive error messages for different API failure scenarios (401, 403, 404, etc.) when fetching models
- API key is not mandatory for fetching models (Openrouter and HuggingFace)

https://github.com/user-attachments/assets/edf33f4f-07ca-4261-aca8-81830edf3578

## Fixes Issues

-

## Self Checklist

- [x] Added relevant comments, esp in complex areas

<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Introduces a `ModelCombobox` for selecting models from provider APIs with enhanced error handling and caching, updating the `AddModel` dialog and adding comprehensive tests.
> 
>   - **Behavior**:
>     - Adds `ModelCombobox` component in `ModelCombobox.tsx` for selecting models from a dropdown fetched from `/v1/models` endpoint.
>     - Supports keyboard navigation and displays error messages for API failures (401, 403, 404).
>     - API key is optional for fetching models from Openrouter and HuggingFace.
>   - **Dialog Updates**:
>     - Replaces input with `ModelCombobox` in `AddModel.tsx` for model selection.
>     - Fetches models using `useProviderModels` hook.
>   - **Hooks**:
>     - Introduces `useProviderModels` in `useProviderModels.ts` to fetch and cache models from provider APIs.
>     - Caches models for 5 minutes and handles errors with specific messages.
>   - **Tests**:
>     - Adds tests for `ModelCombobox` in `ModelCombobox.test.tsx`.
>     - Adds tests for `useProviderModels` in `useProviderModels.test.ts`.
>     - Updates `providers.test.ts` to test new error handling in `fetchModelsFromProvider`.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=menloresearch%2Fjan&utm_source=github&utm_medium=referral)<sup> for aa568e62902cdddf00c46a26c36e21a2a26fa63e. You can [customize](https://app.ellipsis.dev/menloresearch/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

louis-menlo commented:
It works great, but I found two edge cases during testing:
1. When I select the Ollama provider without Ollama running, it shows a refresh button in the dialog. I then open Ollama and click the refresh button → the button disappears (no models are downloaded). After downloading a model from Ollama, the dropdown still shows an empty state and there is no refresh button.
Expected behavior: the refresh button should always be visible.
<img width="1286" height="706" alt="CleanShot 2025-09-02 at 12 04 39@2x" src="https://github.com/user-attachments/assets/52296f04-8d83-4e7e-8205-9f2f747959d0" />

2. Following the same flow as above, but without opening Ollama. Instead, I switch to the Gemini provider and then go back.  It shows an error message for Gemini even though I had selected the Ollama provider.
<img width="1492" height="1058" alt="CleanShot 2025-09-02 at 12 05 01@2x" src="https://github.com/user-attachments/assets/8e0820f1-fccc-4adf-a83c-9065980b2513" />

lugnicca commented:
> It works great, but I found two edge cases during testing:
> 
>     1. When I select the Ollama provider without Ollama running, it shows a refresh button in the dialog. I then open Ollama and click the refresh button → the button disappears (no models are downloaded). After downloading a model from Ollama, the dropdown still shows an empty state and there is no refresh button.
>        Expected behavior: the refresh button should always be visible.
> 
> 
> <img alt="CleanShot 2025-09-02 at 12 04 39@2x" width="1286" height="706" src="https://private-user-images.githubusercontent.com/133622055/484401641-52296f04-8d83-4e7e-8205-9f2f747959d0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY4Mjc5NzYsIm5iZiI6MTc1NjgyNzY3NiwicGF0aCI6Ii8xMzM2MjIwNTUvNDg0NDAxNjQxLTUyMjk2ZjA0LThkODMtNGU3ZS04MjA1LTlmMmY3NDc5NTlkMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwMlQxNTQxMTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZjc0ZTBlN2ViNjlhNzkxZTJiOGIyYTJjZTNlMmE1M2NiMzJmMTg5YjRmODAwZjBiNmYzNDU5NzkzMjcyNTczJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.SFfkUWgWeg5doCUtvMV5ly68N8n4IhOJh0qtvxiKg84">
> 
>     2. Following the same flow as above, but without opening Ollama. Instead, I switch to the Gemini provider and then go back.  It shows an error message for Gemini even though I had selected the Ollama provider.
> 
> 
> <img alt="CleanShot 2025-09-02 at 12 05 01@2x" width="1492" height="1058" src="https://private-user-images.githubusercontent.com/133622055/484401761-8e0820f1-fccc-4adf-a83c-9065980b2513.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY4Mjc5NzYsIm5iZiI6MTc1NjgyNzY3NiwicGF0aCI6Ii8xMzM2MjIwNTUvNDg0NDAxNzYxLThlMDgyMGYxLWZjY2MtNGFkZi1hODNjLTkwNjU5ODBiMjUxMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwMlQxNTQxMTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yOTI2NWM3MzIyNmRmZWQ1YzJiNzczZjQwOWJiNjQ0OGJlMDc1NDMwNzVmYjJlNWU2MjQyY2Q1ZGJhYWVlNTMwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.IFmXRrB6ND147glmYZ9u9-ybdYALYPgCyCVRtTt98s4">

Thanks for the feeback ! 

For the first issue, I added the refresh button in the input instead of inside the error message, useful for refreshing even when the cache is in place, and it covers the case for no models and errors. Let me know what you think !

<img width="838" height="387" alt="image" src="https://github.com/user-attachments/assets/7d720d32-61d6-4d8a-927e-5d9b35cece3d" />

For the second one, I managed to recreate the bug but with Ollama error showing on Gemini provider, it seems it was some kind of race condition that was showing the stale response on the UI for the wrong provider. 
I added a simple ID guard that tracks the current request and ignores responses that don't match, it seems to have done the trick as I don't encounter the case anymore, let me know how it goes on your end !

Thanks again :)

louis-menlo commented:
Ah yes, sorry @lucido-simon

louis-menlo reviewed (approved):
LGTM

louis-menlo commented:
Different behaviors for refreshing the model list

1. Can successfully get the model list from the provider screen
<img width="2272" height="1826" alt="CleanShot 2025-09-08 at 13 07 15@2x" src="https://github.com/user-attachments/assets/19dacf7a-4fac-4770-80cd-1da765ac409f" />

2. Can't get model list from model dropdown
<img width="2272" height="1826" alt="CleanShot 2025-09-08 at 13 07 10@2x" src="https://github.com/user-attachments/assets/d2ef34ef-94ec-4dd0-9200-779523a47414" />

louis-menlo commented:
Awesome, thanks @lugnicca
